### ISSUE TITLE
More internal state for parser

### ISSUE THREAD

**Issue by gagern** [Khan/KaTeX#266]
I'm thinking of modifying the parser so it has more internal state and less method arguments. In particular, the `pos` and `mode` arguments to the methods should be replaced by `this.currentPos` and `this.mode`. I'd also introduce `this.peek` or `this.nextToken` to describe the lexer result for `currentPos`. We could have a new method `consumeToken` or similar to update the current position to the end of the next token, then lex the next token at that new position.

Here are the benefits I envision:
* **More efficiency.** We'd avoid repeatedly lexing at the same position, which is what we currently do as @xymostech confirmed in https://github.com/Khan/KaTeX/pull/246#issuecomment-112604386.
* **Easier communication with environments.** Environments, which call to the parser to read their body, could simply say “hey, parser, give me the next expression”, instead of “hey, parser, give me the expression at the location you gave me and in the mode you gave me”.
* **Ensure linear operation.** If the normal mode of operation is reading one token after the other, then I believe (although this is a bit of a gut feeling) that it might help in implementing macros (#250). We could just make sure that the next few tokens we read are those coming from the definition of the macro. Doing so might mean modifying the `consumeToken` method or changing the lexer, and if you change the lexer, then the `pos` alone isn't sufficient to locate a token any more. I feel (still a gut feeling) that it also might help in annotating each parse node with a begin and end position, which in turn will help getting rid of the `positions` array for function invocations (#255).

Considering possible downsides, I believe the most likely downside is that changing the mode now means ensuring that you change it back again as well. Since the only place where we change the mode is in `parseSpecialGroup`, I think it should be possible to get that right in that one location.

Would you be interested in a modification along these lines? If so, I'll write that, but I'd like to hear whether it has a chance of getting merged before I start on this endeavor.

**Comment by kevinb7** [comment]
> Environments, which call to the parser to read their body, could simply say “hey, parser, give me the next expression”, instead of “hey, parser, give me the expression at the location you gave me and in the mode you gave me”.

I'm all for simplicity and "get the next expression" is definitely simpler.

> Considering possible downsides, I believe the most likely downside is that changing the mode now means ensuring that you change it back again as well.  Since the only place where we change the mode is in parseSpecialGroup, I think it should be possible to get that right in that one location.

That would be my concern as well.  Good to know we only do that in one location.

> If the normal mode of operation is reading one token after the other, then I believe (although this is a bit of a gut feeling) that it might help in implementing macros (#250).

I think in general having simpler methods such as ```nextExpression``` will help implementing future features.

> Would you be interested in a modification along these lines? If so, I'll write that, but I'd like to hear whether it has a chance of getting merged before I start on this endeavor.

I think we should park this one for now.  I'd like to reduce the number of pull requests we have in flight.  I'm afraid that if we add more pull requests we'll end with more stale pull requests.

**Comment by gagern** [comment]
> I'd like to reduce the number of pull requests we have in flight.

Yes, we have quite a few of them around, and I keep adding to them… On the other hand, I'll need macros (per site, defined in JavaScript as part of the options object) pretty soon, and as I said, I think that should build on this simplification here. So I assume that I *will* implement this soon. Unless someone else does macros in a different way before I get a chance. Would you want me to not post a separate pull request for this, but keep it private until I can post a combined pull request for this and macros?

**Comment by kevinb7** [comment]
> Yes, we have quite a few of them around, and I keep adding to them…

The problem with this is that we have limited resources when it comes to reviewing them. @xymostech, @spicyj and myself are all busy working on other projects at the moment so we don't have a lot of time to review lots of pull requests.

> On the other hand, I'll need macros (per site, defined in JavaScript as part of the options object) pretty soon, and as I said, I think that should build on this simplification here.

You've proposed a couple of other refactorings in addition to this.   Unfortunately we won't be able to get all of that reviewed and merged "pretty soon".  Are any of the refactorings necessary to implement macros?  If there isn't then I would suggest implementing macros first.  If not, then let's try to do the minimal refactoring in the short term.  In long term I would like to see more of these refactorings happen.

> Would you want me to not post a separate pull request for this, but keep it private until I can post a combined pull request for this and macros?

In general we'd like to go in the other direction when it comes to pull requests.  I talked with @xymostech about https://github.com/Khan/KaTeX/pull/262 and we'd prefer if that was split up into separate reviews for functions, environments, and symbols.  The smaller the pull request, the easier it is to review and the faster it can get merged.

Before doing that though, please think about whether any of those changes are necessary for macros.  Lastly, I'd like to merge https://github.com/Khan/KaTeX/pull/132 before merging refactorings.  I'm going to finish it up this weekend, but it will still take some time to review.

**Comment by gagern** [comment]
> we have limited resources

Quite understandable, who doesn't.

> You've proposed a couple of other refactorings in addition to this. Unfortunately we won't be able to get all of that reviewed and merged "pretty soon".

Well, if I have macros in a local fork for a while, that should be fine as well, as long as I *do* have them. Reviewing and merging should be a long term goal, but since I already require the render-to-canvas stuff, I'll be on a local fork for the forseeable future in any case.

> Are any of the refactorings necessary to implement macros?

I guess not as such, but the refactorings make it much easier for me to think about this and come up with a clean solution. If you were to say that you don't want any such refactorings at all, I'd try to hack my way around, but I feel that the quality of the solution will be better if I can build on the refactorings. Since my priority is having a maintainable solution for our project, and merging back to KaTeX comes second, I'd go for the clean solution even if getting that back in will take longer.

>  we'd prefer if that was split up into separate reviews for functions, environments, and symbols.

I think first you'd want to split between #255 and #256, which means looking at each commit independently. Then splitting #256 by unit would mean looking at files independently, with the changes to `Parser.js` tied in with those to `functions.js`. If you want to, you can close that request now, and I can open split ones once #132 has landed. The closed request should still be sufficient to illustrate the ideas it was meant to demonstrate.

You should however be aware that the commits from #262 build on one another, and would conflict if parallelized. So my subsequent pull requests would have to be linearized, one after the other, so when you find the time to review a bulk of stuff you might run out of pull requests after the first approved merge. And you should be aware that splitting #256 by unit would temporarily result in inconsistent style.

**Comment by kevinb7** [comment]
> Well, if I have macros in a local fork for a while, that should be fine as well, as long as I do have them. Reviewing and merging should be a long term goal, but since I already require the render-to-canvas stuff, I'll be on a local fork for the forseeable future in any case.

That makes sense.  I'll probably set up a "canvas" branch to cherry pick changes.

> Since my priority is having a maintainable solution for our project, and merging back to KaTeX comes second, I'd go for the clean solution even if getting that back in will take longer.

Sounds like a plan.  There's not rush on our end.  Your addition of environments have really kicked started things again.  We want to continue the momentum at a pace we can sustain.

The progression you've suggested makes sense.  @xymostech set up a trello board.  I'll add the things that we've talked about here to board and send you a link.

> So my subsequent pull requests would have to be linearized, one after the other, so when you find the time to review a bulk of stuff you might run out of pull requests after the first approved merge. And you should be aware that splitting #256 by unit would temporarily result in inconsistent style.

Yeah, I'll review them in sequence as they come in.  I can put them on a separate branch before merging them to master if that helps.

**Comment by kevinb7** [comment]
@gagern here's a link the the trello board: https://trello.com/b/P5DWbSuR/katex-roadmap.

**Comment by gagern** [comment]
> I can put them on a separate branch before merging them to master if that helps.

Right now I see little benefit of having things in your custom branch instead of mine. But thanks for the offer.

**Comment by gagern** [comment]
The core functionality for this is now in place. See https://github.com/gagern/KaTeX/compare/e27559eb82d8c50b430857ff9fe33f6f56cbb06e...gagern:parserState for a preview. Since I've built on some unmerged stuff, and will need to rebase this eventually, I'll not request a pull just now. But if you want to have a peek and tell me your thoughts, feel free to comment here.

**Comment by kevinb7** [comment]
@gagern Thanks for providing a preview.  I probably won't have time until Wednesday to give this a serious look.

**Comment by gagern** [comment]
> here's a link the the trello board…

I've never used Trello before. So far it looks like just one more place where we can comment on these various issues, adding to the confusion. Would you gain anything if I were to sign up for Trello, in terms of coordination? If not, I'm happy enough following as a public reader, posting on GH instead.

**Comment by kevinb7** [comment]
The main benefit is being able to sequence work and tell what's being worked on with a bird's eye view.  Zenhub provides the same thing but within github thus avoid synchronization issues.  Unfortunately, it can only be used with Chrome b/c it requires a Chrome extension/plugin that's why we're using trello and putting up with the overhead of syncing.  Anyways, here's what the board looks like right now:

![screen shot 2015-06-29 at 12 44 44 pm](https://cloud.githubusercontent.com/assets/1044413/8415543/b0088498-1e5c-11e5-922c-2147d930b3d3.png)

I'd like to try to get the number of issues down to a more manageable level (1 page), but in the meantime I think this is helpful at least for me to get the big picture.  I don't think there's going to be all that much discussion on trello.  Most of the comments just link back to github issues or pull requests.

**Comment by gagern** [comment]
I can read it all right. If it's just the sequencing, reading should suffice. Thanks.

**Comment by gagern** [comment]
I noticed that the code in my preview will currently fail to parse some things because the look-ahead token gets lexed in wrong mode. Example: `\text{sin}^2` The `}` to end the group gets consumed and the next token is lexed before the mode returns to math. I see several possible ways of dealing with this:

1. Change the lexer so it lexes the same set of tokens for math and text mode. Catch tokens which are invalid for the current mode later on, when we fail to find a function or symbol for them.
2. Instead of throwing an exception, let the lexer return an error token. Then handle that in the parser, throwing the exception except for when we are mode-switching.
3. Add some argument (called `nextMode` or similar) to the `parseGroup` and `parseOptionalGroup` methods, so that they can switch the mode before consuming their closing brace resp. their sole unbraced token.  
Not sure how to adapt this to the mode switch *before* a group, though, if that group is a single symbol. Right now we neither have a command which explicitely switches to math mode, nor a symbol which is valid in text mode but not in math mode. So right now the problem of a bad lexer mode at the beginning of the group cannot be demonstrated.

Do you have any preference?

**Comment by kevinb7** [comment]
> Example: ```\text{sin}^2``` The ```}``` to end the group gets consumed and the next token is lexed before the mode returns to math.

It seems like we should switch modes right after we encounter the ```}```.  Maybe line 557 in Parser.js should be ```this.expect(optional ? "]" : "}", false);```?  I'll have to check the code out and play with it a bit before I can give a more definitive answer.

**Comment by kevinb7** [comment]
And line 595 should be ```this.expect("}", false);``` and then consume after ```this.mode = outerMode;``` on line 576.  I just had a quick look so that may be completely wrong.

> Not sure how to adapt this to the mode switch before a group, though, if that group is a single symbol
Is this ```2^\text x``` an example of this or does it need to be an explicit group with 1 character?

**Comment by gagern** [comment]
Note to self: Currently `mathNormals` matches `[!"&-\[\]-~]` and `textNormals` matches ``[!"&-=?-\[\]`-\{\}~]`` so the only characters matches by `mathNormals` but not `textNormals` are `>^_|`. The asymmetry between `<` and `>` is probably an error. I assume that the `)-=` in `/[a-zA-Z0-9`!@*()-=+\[\]'";:?\/.,]/` probably wasn't meant as a range. This doesn't affect any characters other than `<`, though. As you can see, entering `\text{foo<bar}` right now gives a parser error even though the lexer allows that symbol through. So we'd get the same kind of error of we were to allow `>^_|` through as well. The only characters from the range 33 through 126 which are matched in neither mode are `#$%\` which makes a lot of sense, except for `$` which we'll probably want in the parser soon enough.

**Comment by gagern** [comment]
Will continue on this another day.

**Comment by gagern** [comment]
> Is this `2^\text x` an example of this or does it need to be an explicit group with 1 character?

Even `\text x` is an example of late mode switching: the `x` gets parsed first while still in math mode, then again in text mode. But we can't tell, because it's allowed in both modes. There is no way to demonstrate the issue with the current set of lexing rules and the current function signatures.

**Comment by gagern** [comment]
> It seems like we should switch modes right after we encounter the `}` …

What you describe is a bit like my option 3 except you don't consume the token in the group, but in the caller. This is problematic since there are many other methods besides `parseSpecialGroup` which call `parseGroup`. At the very least, we should add a `consume` argument to indicate whether the final token should be consumed, and have that argument default to false like it does for `excpect`. Right now I'm slightly in favor of 1.

**Comment by gagern** [comment]
In https://github.com/gagern/KaTeX/commit/8496a1e40f8dfbfc8dd7d2e7b5096bd5c2b01c02 I went with option 1: I got rid of the distinction between text tokens and math tokens. So far it feels like the right choice.

**Comment by gagern** [comment]
Now that all the dependencies are in place, it's time to pick this one here up again. I've cherry-picked two of the three commits for this and filed pull requests. Once those are landed, I can cherry-pick e98b34ebee30238b108bc60969e3f952fb014b34 on top of that. (One benefit of having the lexer rewrite a44ae341af40a63929b626f7402f707b34158a6a merged before the look-ahead handling is that we avoid a known problem due to the way e98b34ebee30238b108bc60969e3f952fb014b34 does look-ahead.)

**Comment by gagern** [comment]
Two of my three commits on this have been merged by Keving, thanks a lot! The remaining one, e98b34ebee30238b108bc60969e3f952fb014b34, will perhaps take some time to cherry-pick. It interfers quite a bit with the `throwOnError` setting which got introduced in the meantime. And perhaps I should address #297 first, to ensure that all the error messages retain their meaning.

**Comment by kevinbarabash** [comment]
I think it makes sense to do https://github.com/Khan/KaTeX/issues/297 first.  Hopefully there isn't too many tests to write.

**Comment by gagern** [comment]
I came up with 43 tests in #385, and it took me quite some time (partly because it's rather boring work). Will continue on this one here, and then on towards macros, once that is in place.

**Comment by kevinbarabash** [comment]
@gagern have you added all of the state you wanted to add to the parser?

**Comment by gagern** [comment]
@kevinbarabash Think so. Next one on my agenda when I find the time is changing the position handling so that a ParserError takes a failing token (or compound construct) instead of a position. That's similar to some of the things I did in 75aad6677b719db4f0fe58283cc5485f51fe2057. But it's not really internal state of the parser any more, more like internal state of the token class.

**Comment by kevinbarabash** [comment]
Sounds good.  I had a quick look at https://github.com/Khan/KaTeX/commit/75aad6677b719db4f0fe58283cc5485f51fe2057... moving the color group, size group, etc. into the parser makes a lot of sense.
